### PR TITLE
SEO: Add module settings to WP Admin

### PR DESCRIPTION
**Pull request by samiff**
#### Changes proposed in this Pull Request:

* Make the SEO Tools module settings available in WP Admin; currently WP Admin links to Calypso for these settings.

#### Jetpack product discussion

* Builds upon https://github.com/Automattic/jetpack/pull/17309
* Internal PT: p1HpG7-ba8-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

1. Pull down these changes on a Jetpack connected site.
2. Observe the SEO Tools settings from the Calypso management page at `wordpress.com/marketing/traffic/__SITE_URL__`
3. Make sure the SEO Tools module is turned on `Enable SEO Tools to optimize your site for search engines`
4. Set some custom `Page Title Structure` values for each page-type/section (e.g. Front Page, Archives, etc.); be sure to use some of the buttons to insert the dynamic values (e.g `+ Site Title`, `+ Tagline`, etc.)
5. Navigate to the equivalent WP Admin settings page at `__SITE_URL__/wp-admin/admin.php?page=jetpack#/traffic` 
6. You should see the same values as in Calypso under the `Search engine optimization` section, though placeholders are used here for the dynamic values (e.g. `[site_name]`) in the inputs.
7. You should be able to edit values and save those settings, and see them reflected in Calypso as well.
8. Check the various features such as the `Front Page Meta Description` and live previews.
9. When activating a "conflicting" plugin such as Yoast, the SEO settings will be disabled and a notice shown.

#### Proposed changelog entry for your changes:

SEO Tools: existing settings now available in WP Admin

#### Screenshots:

WP Admin:

Screen Shot 2021-03-23 at 1 12 17 PM

Screen Shot 2021-03-23 at 1 12 57 PM

Original Calypso implementation:

Screen Shot 2021-03-23 at 1 16 33 PM

### CODE REVIEW

**Comment by samiff** [comment]
Some notes as I look through this:

* Calypso main SEO form component: https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/site-settings/seo-settings/form.jsx
  * Seen when at the `wordpress.com/marketing/traffic/__SITE__` route 
* Meta Title Editor: https://github.com/Automattic/wp-calypso/tree/trunk/client/components/seo/meta-title-editor
* Title Format Editor: https://github.com/Automattic/wp-calypso/tree/trunk/client/components/title-format-editor
* Doesn't seem like the packages are published to NPM: https://github.com/Automattic/wp-calypso/tree/trunk/packages

I _think_ we could get away with our `FormTextarea` from `components/forms` for the "Front Page Meta Description" portion. The Calypso `<CountedTextarea>` component does have some extra features for the character count.

Open for discussion is implementing the controls for the "Page Title Structure"; the Meta/Title Editor components in Calypso seem more specific rather than being generic components.

Also I could use some help in accessing the `advanced_seo_front_page_description` and `advanced_seo_title_formats` settings as I don't currently seem them coming down through `props.settings` for the module. I'd think this would involve one of the wp-json API files, but not entirely sure at this point.

**Comment by jeherve** [comment]
> I _think_ we could get away with our `FormTextarea` from `components/forms` for the "Front Page Meta Description" portion. The Calypso `<CountedTextarea>` component does have some extra features for the character count.

Yes, I don't think we need to add a new component for this. The character count can be added below the form manually, kinda like we do for the Publicize and SEO boxes in the block editor:
- 	https://github.com/Automattic/jetpack/blob/ba4566bf2aabb75a257b4eb0655b5bc63adc4f53/projects/plugins/jetpack/extensions/blocks/publicize/form-unwrapped.js#L90-L101
- https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/extensions/blocks/seo/panel.js#L26-L32

> Open for discussion is implementing the controls for the "Page Title Structure"; the Meta/Title Editor components in Calypso seem more specific rather than being generic components.

That's a bit more tricky. Those Calypso components don't seem quite ready to be packaged up and made available via NPM. Part of me wonders if we wouldn't be better off building the interface again. We don't have any components quite like it in the existing React dashboard, but maybe we could leverage core components instead (we can import core components in the React dashboard, like [here](https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/_inc/client/setup-wizard/feature-toggle/index.jsx#L4)). 

The block editor's `FormTokenField` and the tag management interface, although different, could provide us some clues. 

@tyxla I'd be curious to have your take on this. 

> I could use some help in accessing the advanced_seo_front_page_description and advanced_seo_title_formats settings as I don't currently seem them coming down through props.settings for the module. I'd think this would involve one of the wp-json API files, but not entirely sure at this point.

Yes, you will indeed need the following:

1. Allow those options to be updated via the `jetpack/v4/settings` endpoint. This happens here:
https://github.com/Automattic/jetpack/blob/ab9d8efebed3227a64afe94d0aafd49f1d00ac89/projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php#L2248
2. Once you've done that, you'll be able to rely on the `withModuleSettingsFormHelpers` HOC. If you wrap your own component with that, you'll have access to all the functionality from the HOC, such as `getOptionValue`, `isSavingAnyOption`, or `updateFormStateModuleOption`. Check this to find out more:
https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/_inc/client/components/module-settings/with-module-settings-form-helpers.jsx#L13-L20
You can also check how other components use `withModuleSettingsFormHelpers` to get a better idea of how things work. Carousel, for example, may be a good example since it's a module that can be activated / deactivated and has extra options.

**Comment by tyxla** [comment]
> That's a bit more tricky. Those Calypso components don't seem quite ready to be packaged up and made available via NPM. Part of me wonders if we wouldn't be better off building the interface again. We don't have any components quite like it in the existing React dashboard, but maybe we could leverage core components instead (we can import core components in the React dashboard, like here).

> The block editor's FormTokenField and the tag management interface, although different, could provide us some clues.

> @tyxla I'd be curious to have your take on this.

This bit of Calypso I really dislike TBH. It's the last place where we still make use of `draft-js`, which is a hefty library. Ideally, we wouldn't be using it. 

I'd love to see an approach where we expand on the core `FormTokenField` and we allow it to contain open text tokens between the known tokens. It could be a new component. And then I'd love to use that in Calypso, for the SEO fields. think that's the best way to go.

**Comment by samiff** [comment]
@jeherve The front page meta and social previews have been added in. For the social previews, I've added these in a `<FoldableCard>` whereas Calypso was using a modal for the display, I think this works well but let me know what you think.

I also brought in https://github.com/Automattic/social-logos for the social icons, let me know if this is superfluous or if we have suitable icons in the repo that could be used instead.

I'll be taking a look at the `FormTokenField` (page title structures) bit next.

Screen Shot 2021-02-24 at 7 39 43 PM

**Comment by jeherve** [comment]
> For the social previews, I've added these in a <FoldableCard> whereas Calypso was using a modal for the display

This may be a question better answered by one of our designers, but it's worth noting that we could use the `Modal` component from `_inc/client/components/modal` if we needed to implement this in a modal.

I'll add the "Needs Design" label so we can get some ideas :)

**Comment by jeherve** [comment]
> I also brought in Automattic/social-logos for the social icons, let me know if this is superfluous or if we have suitable icons in the repo that could be used instead.

Since we're only using the package for 3 different icons, I'm wondering if we may not be easier to just add the SVGs for the 3 icons directly, instead of relying on an additional package.

**Comment by samiff** [comment]
> I'll add the "Needs Design" label so we can get some ideas :)

Thanks! One thing I liked about using the `<FoldableCard>` instead was it makes it easier to "live preview" your custom front page meta description as you change it, rather than having to click a button to open a modal and see changes.

**Comment by samiff** [comment]
> Since we're only using the package for 3 different icons, I'm wondering if we may not be easier to just add the SVGs for the 3 icons directly, instead of relying on an additional package.

I could definitely bring in the icons manually if you think that'd be better here. As I understand it, one upside of using the Automattic/social-logos package would be that as a logo or branding guidelines change, those changes would automatically happen rather than needing to manually update the svg sources in our code.

**Comment by jeherve** [comment]
> As I understand it, one upside of using the Automattic/social-logos package would be that as a logo or branding guidelines change, those changes would automatically happen rather than needing to manually update the svg sources in our code.

Indeed. Let's use the package for now, I don't think it has too big of an impact on the whole bundle anyway. 👍

**Comment by samiff** [comment]
I'm marking this as Needs Review and would appreciate if anyone can help me with a few additional thoughts.

In `custom-seo-titles.jsx` there are 3x of `error - JSX props should not use arrow functions - react/jsx-no-bind`

https://github.com/Automattic/jetpack/blob/44f9ef42f10a02c3e8f9e6c9e32e8ca99db6f109/projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx#L160

But I'm not sure what a good alternative would be here when wanting to pass args to a callback?

---

In `class.core-rest-api-endpoints.php`:

https://github.com/Automattic/jetpack/blob/44f9ef42f10a02c3e8f9e6c9e32e8ca99db6f109/projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php#L2790

Do we need a validate/sanitize function here and what would that look like?

There are also a few phpcs warnings here that weren't cooperating 🙂

> Array double arrow not aligned correctly; expected 2 space(s) between "'advanced_seo_front_page_description'" and double arrow, but found 1. (WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned)

---

Calypso defines some conflicting SEO plugins here:

https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/seo/index.js#L6-L11

Which is not a 1:1 match with what JP has here:

https://github.com/Automattic/jetpack/blob/10e29c507a92eae5155b996de1aab8c44228dc2c/projects/plugins/jetpack/modules/seo-tools.php#L18-L27

I'm not quite sure on these, but for now I've added what Calypso has in regards to disabling the seo-tools module when one of the conflicting plugins is active: https://github.com/Automattic/jetpack/pull/18810/commits/44f9ef42f10a02c3e8f9e6c9e32e8ca99db6f109

File: projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-site-settings-endpoint.php
Comment (jeherve) [inline projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-site-settings-endpoint.php:820]: Maybe this could happen directly within `Jetpack_SEO_Titles::update_title_formats` and not as a new method?

**Review by jeherve — COMMENTED** [review]
Sorry for the late reply, I hadn't had the time to come back to this until now :\

> I'm not sure what a good alternative would be here when wanting to pass args to a callback?

Would using `useCallback` work here?

> There are also a few phpcs warnings here that weren't cooperating 🙂

Yeah, I believe you'll need to re-align every item in the array to fix this. Annoying, but  ¯\_(ツ)_/¯

> I'm not quite sure on these, but for now I've added what Calypso has in regards to disabling the seo-tools module when one of the conflicting plugins is active

That seems like a good idea. Part of me wonders if we shouldn't try to consolidate things here, and have one list defined in PHP that we would use [here](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/seo-tools.php#L18), and then expose that list via a new v4 endpoint that we'd pull in the React dashboard. It would avoid duplicating things, but that may be overkill though, what do you think?

File: projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-site-settings-endpoint.php
Comment (samiff) [reply projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-site-settings-endpoint.php]: Refactored for the wpcom side a bit in https://github.com/Automattic/jetpack/pull/18810/commits/fe10ab9e01bc1a22250eff918c26228c3be03124 but the separate `sanitize_title_formats` function is needed to serve the wp-admin side which doesn't call `update_title_formats`

**Comment by samiff** [comment]
> have one list defined in PHP that we would use [here](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/seo-tools.php#L18), and then expose that list via a new v4 endpoint

@jeherve Thanks for the comments 👍   The only issue I'm still unclear on is the one for conflicting SEO plugins. I think it would make sense to have a single source of truth at least on the JP side of things. Though I'm not all that familiar with the REST API setup, would you be able to give an overview of what might be needed here if we want to do that?

On the other hand, those plugins aren't changed all too often it looks like, so maybe just some good code comments would suffice?

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/traffic/seo.jsx:50]: Maybe we can remove this now?

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/traffic/seo.jsx:90]: Would it make sense to import `fetchPluginsData`, `isFetchingPluginsData`, and `isPluginActive` from `state/site/plugins` right here, and rely on `isPluginActive` to check for each conflicting plugin?

File: projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx:129]: I think it may be useful to have this being one string with `value` as a placeholder.

File: projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx:161]: How about using `useCallback` directly for `handleTokenButtonClick` here as well, and keep everything inside the `CustomSeoTitles` component?

Here is an example that may help:
https://github.com/Automattic/jetpack/blob/6c7323b1ecb76efda265aa58f4a3a7c95b8cad7f/projects/plugins/jetpack/_inc/client/recommendations/feature-summary/index.jsx#L41-L45

**Review by jeherve — COMMENTED** [review]
It's some really nice work so far 👏 , it's coming together nicely. 

> The only issue I'm still unclear on is the one for conflicting SEO plugins. I think it would make sense to have a single source of truth at least on the JP side of things. Though I'm not all that familiar with the REST API setup, would you be able to give an overview of what might be needed here if we want to do that?

We can take a look at that in a follow-up PR, I think we could ship a v1 without this, it's not that important imo.

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo.jsx]: Good catch! https://github.com/Automattic/jetpack/pull/18810/commits/5debeb6cb1558aca79306b2b5eb6cd353ea037a1

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo.jsx]: Nice, didn't realize we had those functions: https://github.com/Automattic/jetpack/pull/18810/commits/2a06b19e6d6c96b16f1b3eb89d2cd31c6f409f2e

File: projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx]: Agreed, https://github.com/Automattic/jetpack/pull/18810/commits/abfc9a00b7e4c7047401a26b42537f782fe3ae79

File: projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx]: I was able to refactor a bit and leverage `useCallback` for the click handler https://github.com/Automattic/jetpack/pull/18810/commits/bef5e3f7277e59c53ef6ab0967483e0afca8c35c

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/traffic/seo.jsx:166]: I wonder if this one (and the one below) should also be `FoldableCard`, hidden by default, to avoid making the settings page too long?

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/traffic/seo.jsx:184]:
```suggestion
											'Craft a description of your Website: up to 160 characters that will be used in search engine results for your front page, and when your website is shared on social media sites.',
```

**Review by jeherve — COMMENTED** [review]
This works well for me. 

The more I play with this, the more I wonder if it may make sense to move away from an `input` field for each custom input, and instead have a `contenteditable` div where you'd be able to insert the same content as you've built already. Maybe that would allow us to  have non-editable tokens inside that div, instead of the current `[something]` strings. What do you think?

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo.jsx]: Yes, I think it's a good idea to add them as foldable cards https://github.com/Automattic/jetpack/pull/18810/commits/87d76d81562c2ad6beee49a5328dcd4c88044927

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo.jsx]: https://github.com/Automattic/jetpack/pull/18810/commits/ea01209539e06d732fa9165a950018881c49928d

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (samiff) [reply projects/plugins/jetpack/_inc/client/traffic/seo.jsx:27]: The names are used to output the conflicting plugin name to the user:

https://github.com/Automattic/jetpack/blob/ea01209539e06d732fa9165a950018881c49928d/projects/plugins/jetpack/_inc/client/traffic/seo.jsx#L137

**Comment by samiff** [comment]
> The more I play with this, the more I wonder if it may make sense to move away from an `input` field for each custom input, and instead have a `contenteditable` div where you'd be able to insert the same content as you've built already. Maybe that would allow us to have non-editable tokens inside that div, instead of the current `[something]` strings. What do you think?

@jeherve  Thanks for taking a look 👍  About the `contenteditable`, I did consider that initially but it seemed like there were several gotcha cases (pasting as html, line breaks, cursor position, etc.) to watch for and it wasn't quite as simple as one would think to manage the tokens within. If you have more experience with that I'd defer to you in gauging how difficult building out a component for this might be.

**Review by jeherve — APPROVED** [review]
This looks good to me. I think we should be able to merge this now. UX can always be improved (especially around the display of the tokens as we discussed above), but I think we're in a good place already, and can iterate in future PRs. 

Great job on this @samiff! 🎉 

Let's discuss the status of the module needing a connection to wpcom in a follow-up issue / PR maybe?

File: projects/plugins/jetpack/_inc/client/traffic/seo.jsx
Comment (sergeymitr) [reply projects/plugins/jetpack/_inc/client/traffic/seo.jsx:122]: We've discussed that a bit here: p1HpG7-ba8-p2#comment-44940
The plan is to make it userless as soon as this PR gets merged.

**Comment by samiff** [comment]
r223439-wpcom

**Pull request merged**
